<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<title>Untitled Document</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
</head>

<body>
<CENTER>
  <H1><EM>Data Repository</EM></H1>
  <em><strong>SRS Document</strong></em><BR>
  <EM>Sam Angiuoli</EM><BR>
  <EM>TIGR</EM><BR>
  <EM>$Date$</EM><BR>
  <EM>$Revision$</EM> 
</CENTER>
<H2><a name="a1"></a>1. Introduction</H2>
<UL>
  <P> 
  <LI><B><a name="a1b1"></a>1.1 Purpose of this document</B><BR>
    This document describes the requirements, directory structure, management methods, and access methods for a repository 
    of data files. 
    <P></P>
  <LI><B><a name="a1b2"></a>1.2 Scope of this document</B><BR>
    The data repository in this document will be used by the data management scripts 
    and workflows of the <a href='/ifxdevel/papyrus'>Papyrus</a> software package. 
    The repository may hold any data files associated with the Papyrus pipelines. 
	Each project will have its own data repository.  A project may contain data for a single genome or a set of genomes.
    <P></P>
  <LI><B><a name="a1b3"></a>1.3 Overview</B><BR>
    <ul>
	<li>The data repository must consist of a well-defined set of directories</li>
	<li>Authentication to read/write files to the repository is not required</li>
	<li>Software is responsible for handling synchronization and concurency issues</li>
	<li>The data repository must be backed up and recoverable</li> 
	<li>Local access to the repository via a unix mount point is required.  This includes NFS mounts.</li>
	<li>Remote access to the repository via a proxy server (such as WebDAV) is optional.  This may be a requirement in the future.</li>
	</ul>
    <P></P>
  <LI><B><a name="a1b4"></a>1.4 Business Context</B><BR>
    The data repository is the primary data store for large amount of data generated 
    during a pipeline. Effective management and organization of this data is critical.</LI>
</UL>
<H2><a name="a1"></a>2. Repository directory structure</H2>
The data repository lives under a root directory, referred to in this document as <em>REPOSITORY_ROOT</em>.  
Software that accesses the repository should be able to configure the value of <em>REPOSITORY_ROOT</em>.
<ul>
<li><b>2.1 Top level directories</b></li>
<pre>
<em>REPOSITORY_ROOT</em>/BSML_repository
<em>REPOSITORY_ROOT</em>/FASTA_repository
<em>REPOSITORY_ROOT</em>/WORKFLOW_repository
</pre>
<p></p>
<li><b>2.2 BSML_repository</b></li><br>
<p></p>
The BSML_repository contains XML files conforming to the <a href='/ifxdevel/cas/bsml'>BSML</a> DTD.  
The top level of the BSML_repository (<em>REPOSITORY_ROOT</em>/BSML_repository/*.bsml) contains files 
representing genomic sequences, biological features (eg. genes), and associated annotations.  
These files are assumed to be static for their lifetime in the repository and should not be generated as part of a computational pipeline.  
These files may be generated from a database.  These files provide "primary data" required for downstream pipelines.  
The required content ("primary data") in these files will change to meet the requirements of downstream computational pipelines.  
Validation of data contained in these files will be performed using XML schema (see the <a href='/ifxdevel/cas/bsml'>check_repository</a> utility for more information about validation).
<p></p>
<li><b>2.2.1 BSML_repository sub directories</b></li>
<p></p>
BSML files derived from compuational pipelines are stored in sub directories below BSML_repository.  
All sub directories in the BSML_repository are optional and creation of these directories are handled by steps in the workflow.
The name of the sub directory should match the name of the workflow used to generate the data.  
<pre>
<em>REPOSITORY_ROOT</em>/BSML_repository/nucmer
<em>REPOSITORY_ROOT</em>/BSML_repository/promer
<em>REPOSITORY_ROOT</em>/BSML_repository/blastp
</pre>
<p></p>
<li><b>2.2.1 BSML_repository checksums</b></li><br>
A file named 'CHECKSUM' exists in all directories in the BSML_repository including the top level directory.  
This file contains the md5 checksum for each BSML file in the direcotory.  
In addition, a file named 'CHECKSUM.SUM' exists in all directories in the BSML_repository.   
The CHECKSUM.SUM file contains the md5sum of the CHECKSUM file.  
The single value in the CHECKSUM.SUM file represent a checksum on all the data at a particular level of the repository.  
This value can be thought of as a checksum for an entire data set as opposed to a checksum for an individual file.
<p></p>
<li><b>2.3 FASTA_repository</b></li><br>
The FASTA_repository should contain files in FASTA format in the top level directory.  These files are assumed to be static for their lifetime in the repository and should not be generated as part of a computational pipeline.  
These files may be generated from a database.  These files provide FASTA sequences required for downstream pipelines.  
The required sequences in these files will change to meet the requirements of downstream computational pipelines.  
Validation of sequences contained in these files will be performed using the <a href='/ifxdevel/cas/bsml'>check_repository</a> utility.
<pre>
<em>REPOSITORY_ROOT</em>/FASTA_repository
</pre>
<p></p>
<li><b>2.3.1 FASTA_repository sub directories</b></li><br>
No sub directories are planned at the moment.  This may change with requirements for the computational pipelines.
<p></p>
<li><b>2.3.1 FASTA_repository checksums</b></li><br>
A file named 'CHECKSUM' exists in all directories in the FASTA_repository including the top level directory.  
This file contains the md5 checksum for each FASTA file in the direcotory.  
In addition, a file named 'CHECKSUM.SUM' exists in all directories in the FASTA_repository.   
The CHECKSUM.SUM file contains the md5sum of the CHECKSUM file.  
The single value in the CHECKSUM.SUM file represent a checksum on all the data at a particular level of the repository.  
This value can be thought of as a checksum for an entire data set as opposed to a checksum for an individual file.
<p></p>
<li><b>2.4 WORKFLOW_repository</b></li><br>
The top level of the WORKFLOW_repository should not contain any files.
<pre>
<em>REPOSITORY_ROOT</em>/WORKFLOW_repository
</pre>
<p></p>
<li><b>2.4.1 WORKFLOW_repository sub directories</b></li><br>
All sub directories in the WORKFLOW_repository are optional and creation of these directories are handled by the wrapper scripts that spawn workflow (see <a href='/ifxdevel/cas/papyrus/'>Papyrus</a> for more information).
The name of the sub directory should match the name of the workflow.   
The workflow UID is used as a sub directory below the workflow name.  
This direcotry should contain all the files associated with a particular instance of workflow.  
These files include the template XML file, INI file, XML instance file, workflow log files, all log files produced steps in the workflow.
<pre>
<em>REPOSITORY_ROOT</em>/WORKFLOW_repository/blastp/17893
<em>REPOSITORY_ROOT</em>/WORKFLOW_repository/blastp/17897
<em>REPOSITORY_ROOT</em>/WORKFLOW_repository/blastp/17899
<em>REPOSITORY_ROOT</em>/WORKFLOW_repository/nucmer/17893
</pre>

<H2><a name="a1"></a>3. Repository access</H2>
The <em>REPOSITORY_ROOT</em> must be accessible via a UNIX file system mount point.  
Read/writes to the repository are performed using standard UNIX system calls for IO.

<H2><a name="a1"></a>4. Limitations</H2>
The current specification for the data repository contains many limitations.  
<ul>
<li>Synchronization</li><br>
Writes to the repository are not synchronized.  There is no procedure in place to prevent race conditions occuring during reading and writing files.
<li>Authentication/Perimissions</li><br>
Access to the repository is not authenticated.  Access permissions are regulated by the UNIX file system permissions.  
This can be problematic in a multi-user environment where several users need to read/write files to the repository.
Currently, all files in the repository are world readable/writable to avoid problems.  
It is the responsiblity of steps in a particular workflow to set the file permissions correctly.
<li>Remote(outside) access</li><br>
Remote access to the repository is not possible, at least not through any secure channell.  
Support for remote repositories would expand the capabilities of the compuational pipelines.
</ul>
<p></p>
<H2><a name="a1"></a>5. Enhancements</H2>
This section describes possible enhancements and future efforts.
<H2><a name="a1"></a>5.1 WebDAV</H2>
<a href='http://www.webdav.org'>WebDAV</a> should be explored as a possible solution to some of the limitations of the current implementation.
The performance and scalability of a WebDAV server needs to be evaluated.  
The repository is expected to contain large files and accessing files through a WebDAV server may introduce a bottleneck into the computational pipelines.
The WebDAV protocol appears to solve many limitations of the current implementation.
<ul>
<li>Concurrency/synchronization</li><br>
A repository can be locked using WebDAV, enabling synchronized writes and avoiding race-conditions during reads/writes.
<li>Authentication</li><br>
Authentication is available via HTTPS.
<li>Permissions</li><br>
All files are written by the WebDAV server UNIX file permission problems.
<li>Remote access</li><br>
The WebDAV protocol runs over HTTP 
</ul> 
</ul>
  <P> 
</body>
</html>
