<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">

<html>

<head>
    <meta http-equiv="Content-Language" content="en-us">
    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
    <title>ergatis</title>
    <link rel="stylesheet" type="text/css" href="css/common.css">
    <link rel="stylesheet" type="text/css" href="css/header.css">
    <link rel="stylesheet" type="text/css" href="css/monitor.css">
    <link rel="stylesheet" type="text/css" href="css/sourceforge.css">
</head>

<body class='documentation'>

<div id='page_container'>

    <div id='page_header'>
        <div id='quick_links_container'>
            <ul id='quick_links'>
                <li><a href='http://www.tigr.org'>tigr</a></li>
                <li class='last'><a href='http://www.sf.net/projects/ergatis'>SF site</a></li>
            </ul>
        </div>
        <div id='title_container'>
            <span id='page_title'>ergatis</span>
            <span id='page_slogan'>workflow creation and monitoring interface</span>
        </div>
    </div>

    <div id='primary_nav_container'>
        <ul id='primary_nav_elements'>
            <li id='nav_home'><a href="index.html">home</a></li>
            <li id='nav_screenshots'><a href="screenshots.html">screenshots</a></li>
            <li id='nav_documentation'><a href="documentation.html">documentation</a></li>
            <li id='nav_download'><a href="download.html">download</a></li>
        </ul>
    </div>

    <div id='sub_nav_container'>
        <ul id='sub_nav_elements'>
            <li class='last'><a href='index.html'>home</a></li>
        </ul>
    </div>

    <div id='content_container'>
    
        <div id='section1'>
            <h2>selected component list</h2>
            <ul id='component_list'>
                <li>AAT (aa)</li>
                <li>AAT (na)</li>
                <li>augustus</li>
                <li>backupdb</li>
                <li>BER</li>
                <li>blast (see ncbi-blast and wu-blast series)</li>
                <li>bsml -&gt; chado (db loader)</li>
                <li>bsml -&gt; fasta</li>
                <li>build supercontig</li>
                <li>chado -&gt; bsml (dumper)</li>
                <li>clustalw</li>
                <li>cogs</li>
                <li>coordinate adjustment</li>
                <li>copydb</li>
                <li>coverage</li>
                <li>delete analysis (db manip)</li>
                <li>fasta -&gt; bsml</li>
                <li>fgenesh</li>
                <li>genbank -&gt; bsml</li>
                <li>genemark</li>
                <li>genewise</li>
                <li>genezilla</li>
                <li>genie</li>
                <li>genscan</li>
                <li>genscan+</li>
                <li>gff -&gt; bsml</li>
                <li>glimmer3</li>
                <li>glimmerHMM</li>
                <li>hmmpfam</li>
                <li>hmmpir</li>
                <li>hmmsearch</li>
                <li>hmmsmart</li>
                <li>hmmtigr</li>
                <li>infernal</li>
                <li>initdb (chado db init)</li>
                <li>iprscan</li>
                <li>jaccard</li>
                <li>legacy -&gt; bsml</li>
                <li>mask analysis</li>
                <li>mepipe</li>
                <li>mreps</li>
                <li>muscle</li>
                <li>ncbi-blastn</li>
                <li>ncbi-blastp</li>
                <li>ncbi-blastx</li>
                <li>ncbi-tblastn</li>
                <li>ncbi-tblastx</li>
                <li>ngap</li>
                <li>nucmer</li>
                <li>nucmer SNP detection</li>
                <li>obo -&gt; chado</li>
                <li>pangenome</li>
                <li>pe</li>
                <li>pepstats</li>
                <li>phat</li>
                <li>predotar</li>
                <li>promer</li>
                <li>prosite scan</li>
                <li>psipred</li>
                <li>region</li>
                <li>repeatmasker</li>
                <li>signalp</li>
                <li>snoscan</li>
                <li>snp (detection)</li>
                <li>split fasta</li>
                <li>split multifasta</li>
                <li>statistics -&gt; chado</li>
                <li>subfeature mapping</li>
                <li>tRNAscan-SE</li>
                <li>targetp</li>
                <li>tiling</li>
                <li>tmhmm</li>
                <li>translate sequence</li>
                <li>transterm</li>
                <li>trf</li>
                <li>wait (test component)</li>
                <li>wu-blastn</li>
                <li>wu-blastp</li>
                <li>wu-blastx</li>
                <li>wu-tblastn</li>
                <li>wu-tblastx</li>
                <li>xdformat</li>
            </ul>
        </div>

        <div id='section2'>
            <h2>documentation</h2>
            <p>
                I'm just throwing some notes here to get an idea of the sort of documents I want to be publicly
                available.  Documentation is currently either being written or decoupled from internal TIGR
                specifications (much like the rest of the software.)
            </p>
            <dl>
                <dt><a>overview</a></dt>
                <dd>
                    A description of the overall usage and goals of ergatis will go here.  This should include
                    comparisons with other data management packages.
                </dd>
                <dt><a>component documentation</a></dt>
                <dd>
                    A 'component' in ergatis is a tool, script or series of comments that performs a distinct
                    computational task.  Examples include format conversion, sequence alignment (BLAST, AAT),
                    or gene prediction programs (augustus, glimmerHMM).
                </dd>
                <dt><a>component creation tutorial</a></dt>
                <dd>
                    Ergatis currently contains 93 components and more are added every month.  If there is a specific
                    analysis tool you want to use in the system that doesn't exist as a component you can request
                    it's support or write the component yourself.  A tutorial on how this is done will appear here. 
                </dd>

                <dt><a>directory structure convention</a></dt>
                <dd>
                    Ergatis creates a lot of files and folders when used, including all output result files.  This 
                    documentation should describe the standard directory structures employeed.
                </dd>
                <dt><a>pipeline creation tutorial</a></dt>
                <dd>
                    This is where stuff actually gets done.  You may have 10,000 sequences you want to BLAST across
                    a compute grid, or a few gene finders to run locally.  This is where you learn how to use
                    the interface to build and run your own custom analysis pipeline.
                </dd>
            </dl>
        </div>


    </div> <!-- end content_container -->

    <div id="footer">
        <div id='official_links'>
            <a href='http://www.sf.net/projects/ergatis'><img src="http://sourceforge.net/sflogo.php?group_id=59541&amp;type=1" alt='sourceforge'></a>
            <a href='http://www.opensource.org/'><img src='http://opensource.org/trademarks/open_source_button.png' alt='open source'></a>
            <a href='http://www.getfirefox.com'><img src='images/official_getfirefox.png' alt='get firefox'></a>
            <a href='http://www.perl.com'><img src='images/official_perl.png' alt='perl'></a>
            <img src='images/official_css2.png' alt='css2'>
        </div>
        <div id='license'>
            this project is licensed under the <a href='http://www.opensource.org/licenses/artistic-license.php'>Artistic License</a>.<br>
            questions should be directed to jorvis AT tigr.org
        </div>
        <div style='clear: both;'>&nbsp;</div>
    </div>


</div> <!-- end page_container -->

</body>

</html>
